pjsip/include/pjsua-lib/pjsua.h

```diff
@@ -5573,6 +5573,40 @@ PJ_DECL(pj_status_t) pjsua_call_hangup(pjsua_call_id call_id,
 				       const pj_str_t *reason,
 				       const pjsua_msg_data *msg_data);
 
+/**
+ * Hangup call by using method that is appropriate according to the
+ * call state. This function is different than answering the call with
+ * 3xx-6xx response (with #pjsua_call_answer()), in that this function
+ * will hangup the call regardless of the state and role of the call,
+ * while #pjsua_call_answer() only works with incoming calls on EARLY
+ * state.
+ *
+ * If parameter immediate is set to PJ_TRUE, call DISCONNECTED state
+ * will be immediately reported to app and the call slot can be reused.
+ * The only exception is if media transport creation is in progress
+ * during call setup. In this case, the hangup will be pending upon
+ * the completion of the media transport.
```

Comment:
For the exception condition, will the app still be notified immediately?

Reply:
No, the app will be notified upon completion of media transport.

Reply:
Why is that?

Not sure if this is relevant. With trickle ICE, media transport completion may be reported immediately, while internally ICE is actually still gathering candidates. So for this case, which completion will the pjsua-lib wait for?

Reply:
It is the current behavior of call hangup, i.e. to delay the hangup until media transport creation completes:
https://github.com/pjsip/pjproject/blob/master/pjsip/src/pjsua-lib/pjsua_call.c#L2760

The reason this behavior is not changed is because of the complexity:
In order to support immediate hangup, we move the current `pjsua_var.call` instance into another variable `pjsua_var.hangup_call` and handle the hangup process from this new variable. For media related callbacks however, since currently we are using `&call.media[i]` as the user data, this will point to the old call media instance. It is possible to overcome this issue, by:
- Creating the user data in the heap (using dialog's pool?), perhaps a struct of { dialog and media index }.
- Modifying `pjsua_media.c` callbacks to use this new user data.

But there is still the complexity of race condition, i.e. since the media transport creation is async, there is a possibility that the media callback will happen at the same time with call hangup. This may require us to hold dialog lock in all media related callbacks. So I'm not sure if the potentially major modifications required is worth the advantage, i.e. to be able to immediately hangup the call when async med tp creation is in progress in the beginning of the call (again, note that sync media reinit in the middle of the call should be unaffected by this).

As for trickle ICE, I believe it will be upon the report of completion, because the hangup will proceed in `pjsua_call.c`'s `on_make/incoming_call_med_tp_complete()`. I'm not sure if this can cause any complications though, since trickle ICE is a new feature.

Reply:
As the ticket is about immediate hangup notification and handle the rest in background, I am thinking that media transport creation completion should be part of things to be done in background.

Re: `&call.media[i]` user data in media callbacks, have not checked the patch details, but perhaps the problem raises because the call instance may now be reused, right? What if it cannot be reused until the hangup process completed, e.g: simply add a flag `hangingup` in `pjsua_call` and alloc_call_id() filters out call slot with that flag set. This may be a completely different approach compared to this current patch (which seems to move the hanging up call instance to `pjsua_var.hangup_call`, cmiiw).

Re: trickle ICE, from PJSUA point of view, create_ice_media_transport() will set `call_med->tp_ready` to `PJ_SUCCESS` (instead of `PJ_EPENDING` as in current behavior). So perhaps PJSUA will try to destroy the `call_med->tp` immediately on call hangup, may be safe though.

Reply:
Alright, so that's another alternative we need to consider, i.e. only to report the DISCONNECTED state immediately while the call instance itself still remains. And this seems a hybrid approach between what the PR proposes and the current state (which is that application is recommended to treat the call as disconnected immediately after hanging up (i.e. report as such to the user, clear the call from the app UI, etc), even though it hasn't received DISCONNECTED call state and the hangup process is still underway).

There may be a problem with the alternative approach, i.e. when the call slots are full because of the pending hangups, but users can't make new calls (returning PJ_ETOOMANY) even though `pjsua_call_get_count()` seems to indicate that there are few/no active calls. Even worse, compared to current situation, the user will no longer be notified when the call will finally be cleared.
A solution for this that I can think of is probably to add a new API `pjsua_call_get_free_count()` and introduce a new state `CALL_STATE_DESTROYED`.

There is also another option that we should consider, which is to stick with the current approach. If both alternatives do not actually give a clear advantage over the current one, and only adds unnecessary complexity, I guess it makes sense to leave it as is?

Reply:
The complexity should be the reason the ticket have not been implemented, yet we can't close the ticket because requests keep coming from time to time :) So let's do this I think.

Re: call slot full in the alternative approach, I guess it is understandable, even it can also happen with `pjsua_var.hangup_call` approach. But yes, app should be able to check for any free call slot, either via notification or function call (not sure why we need a new inv state DESTROYED, better not touching PJSIP level if possible).

Some checklist (for all approach, perhaps already handled by this patch, not checked):
- PJSUA still should invoke callback `on_stream_destroyed()`, app may need it to destroy its custom component for example. So perhaps PJSUA should also deinit media immediately as if call is already disconnected.
- Handle/auto-reject incoming re-INVITE/UPDATE/other request silently (don't invoke PJSUA callback).

Reply:
Okay, I will do this in another branch and create a different PR since it is quite a different solution compared to this PR.

Reply:
A few notes when trying to implement the new approach:
- In general, there's a problem with the requirement in #1049: cease any PJSUA call related callback once call is declared to be disconnected. More about this below.
- `on_stream_destroyed()` will be called after the DISCONNECTED `on_call_state()`. But then, so is the current behaviour. Except that since we now report the DISCONNECTED state much earlier, the delay can be longer between the two callbacks. Trying to call `on_stream_destroyed()` earlier poses two problems:
  * The media will already be deactivated upon call disconnection, resulting in call dump not showing any useful end-of-call statistics. But this is minor and the workaround is probably to do the call dump before the media deinit.
  * The bigger problem is the async media transport creation, so no matter what, we still need to wait for it to complete. 
- How about `on_call_tsx_state()`? In #1627 (Callback on_tsx_state_changed() not invoked on call disconnection), the workaround is to call it before `on_call_state(DISCONNECTED)`, but since we now report the DISCONNECTED state earlier in `pjsua_call_hangup()`, we don't have the `pjsip_event`.

Reply:
> A few notes when trying to implement the new approach:
> 
> * In general, there's a problem with the requirement in #1049: cease any PJSUA call related callback once call is declared to be disconnected. More about this below.
> * `on_stream_destroyed()` will be called after the DISCONNECTED `on_call_state()`. But then, so is the current behaviour. Except that since we now report the DISCONNECTED state much earlier, the delay can be longer between the two callbacks. Trying to call `on_stream_destroyed()` earlier poses two problems:
>   
>   * The media will already be deactivated upon call disconnection, resulting in call dump not showing any useful end-of-call statistics. But this is minor and the workaround is probably to do the call dump before the media deinit.

Ic.

>   * The bigger problem is the async media transport creation, so no matter what, we still need to wait for it to complete.

If `on_stream_created()` is not called yet (currently this should be the case when med tp is not completed yet), we can skip `on_stream_destroyed()' perhaps?

> * How about `on_call_tsx_state()`? In #1627 (Callback on_tsx_state_changed() not invoked on call disconnection), the workaround is to call it before `on_call_state(DISCONNECTED)`, but since we now report the DISCONNECTED state earlier in `pjsua_call_hangup()`, we don't have the `pjsip_event`.

Perhaps we can invoke the callback using user event?

Reply:
> >   * The media will already be deactivated upon call disconnection, resulting in call dump not showing any useful end-of-call statistics. But this is minor and the workaround is probably to do the call dump before the media deinit.

Calling `pjsua_call_dump()` in the application's `on_stream_destroyed()` seems to solve this issue.

> If `on_stream_created()` is not called yet (currently this should be the case when med tp is not completed yet), we can skip `on_stream_destroyed()' perhaps?

Yes.

> > * How about `on_call_tsx_state()`? In #1627 (Callback on_tsx_state_changed() not invoked on call disconnection), the workaround is to call it before `on_call_state(DISCONNECTED)`, but since we now report the DISCONNECTED state earlier in `pjsua_call_hangup()`, we don't have the `pjsip_event`.
> Perhaps we can invoke the callback using user event?

`on_call_tsx_state()` also has `pjsip_transaction *` parameter. Passing NULL may cause application to crash.

Reply:
> `on_call_tsx_state()` also has `pjsip_transaction *` parameter. Passing NULL may cause application to crash.

Ah okay.
For immediate hangup, perhaps we should just skip `on_call_tsx_state()` altogether as any tsx state change, if any, happens in background. And the docs should explicitly specify this.

Reply:
About calling pjsua_call_dump() in on_stream_destroyed().
With this change, if we receive the hang up, the call_dump appears as "DISCONNCTD", but if we are the ones who hang up the call, it appears as "CONFIRMED".
Before, in on_call_state, in all cases the call dump appears as "DISCONNCTD".

Reply:
Will continue this discussion in the relevant PR #2600 instead.